libraries/AP_Soaring/AP_Soaring.h

```diff
@@ -0,0 +1,163 @@
+/* 
+Soaring Controller class by Samuel Tabor
+Provides a layer between the thermal centring algorithm and the main code for managing navigation targets, data logging, tuning parameters, algorithm inputs and eventually other soaring strategies such as speed-to-fly. AP_TECS libary used for reference.
+*/
+#pragma once
+
+#include <AP_AHRS/AP_AHRS.h>
+#include <AP_Param/AP_Param.h>
+#include <DataFlash/DataFlash.h>
+#include <AP_Math/AP_Math.h>
+#include "ExtendedKalmanFilter.h"
+#include <AP_SpdHgtControl/AP_SpdHgtControl.h>
+
+#define EXPECTED_THERMALLING_SINK 0.7
+#define INITIAL_THERMAL_STRENGTH 2.0
+#define INITIAL_THERMAL_RADIUS 30.0 //150.0
+#define INITIAL_STRENGTH_COVARIANCE 0.0049
+#define INITIAL_RADIUS_COVARIANCE 2500.0
+#define INITIAL_POSITION_COVARIANCE 300.0
+#define ASPD_FILT 0.05
+#define TE_FILT 0.03
+#define TE_FILT_DISPLAYED 0.15
+
+class SoaringController
+{
+    ExtendedKalmanFilter _ekf;
+    AP_AHRS &_ahrs;
+    AP_SpdHgtControl *&_spdHgt;
+    const AP_Vehicle::FixedWing &_aparm;
+
+    // store aircraft location at last update
+    struct Location _prev_update_location;
+
+    // store aircraft location at last update
+    struct Location _prev_vario_update_location;
+
+    // store time thermal was entered for hysteresis
+    unsigned long _thermal_start_time_us;
+
+    // store time cruise was entered for hysteresis
+    unsigned long _cruise_start_time_us;
+
+    // store time of last update
+    unsigned long _prev_update_time;
+
+    // store time of last update
+    unsigned long _prev_vario_update_time;
+
+    float _vario_reading;
+    float _filtered_vario_reading;
+    float _last_alt;
+    float _alt;
+    float _last_aspd;
+    float _last_roll;
+    float _last_total_E;
+    bool _new_data;
+    float _loiter_rad;
+    bool _throttle_suppressed;
+    
+    float _aspd_filt;
+    uint8_t _msgid;
+    uint8_t _msgid2;
+    float correct_netto_rate(float climb_rate, float phi, float aspd);
+    float McCready(float alt);
+    void get_wind_corrected_drift(const Location *current_loc, const Vector3f *wind, float *wind_drift_x, float *wind_drift_y, float *dx, float *dy);
+    void get_altitude_wrt_home(float *alt);
+  
+protected:
+    AP_Int8 soar_active;
+    AP_Float thermal_vspeed;
+    AP_Float thermal_q1;
+    AP_Float thermal_q2;
+    AP_Float thermal_r;
+    AP_Float thermal_distance_ahead;
+    AP_Int16 min_thermal_s;
+    AP_Int16 min_cruise_s;
+    AP_Float polar_CD0;
+    AP_Float polar_B;
+    AP_Float polar_K;
+    AP_Float alt_max;
+    AP_Float alt_min;
+    AP_Float alt_cutoff;
+        
+public:
+    SoaringController(AP_AHRS &ahrs, AP_SpdHgtControl *&spdHgt, const AP_Vehicle::FixedWing &parms, uint8_t msgid, uint8_t msgid2) :
+        _ahrs(ahrs),
+        _spdHgt(spdHgt),
+        _aparm(parms),
+        _thermal_start_time_us(0),
```

Comment:
please don't init variables to zero in constructors. Due to a g++ bug it just takes up flash space to no purpose. The variables are guaranteed to be initialised to zero already